### ISSUE TITLE
deno_tcp latency is worse than node_tcp

### ISSUE THREAD

**Issue by ry** [denoland/deno#2700]
<img width="719" alt="Screen Shot 2019-07-31 at 8 09 21 AM" src="https://user-images.githubusercontent.com/80/62210605-8ae9e380-b36a-11e9-95f9-156119a6c488.png">

Thanks to @bartlomieju's work in https://github.com/denoland/deno/pull/2689 we're able to clearly see now that deno_tcp has worse 99% tail latency than node_tcp. 

Previously I believed that our max latency was on par with Node's. [Back in April](https://github.com/denoland/deno/pull/2131), @piscisaureus did work to improve our latency numbers, but apparently it was not enough.

This is a very good place to investigate performance problems. 

Note that deno_core_single does not exhibit the latency problems! deno_core_multi is worse, but not as bad as deno_tcp....

**Comment by bartlomieju** [comment]
Ref #2758

**Comment by bartlomieju** [comment]
Interestingly [deno_http](https://deno.land/std/http/http_bench.ts) has 4x lower latency than [deno_tcp](https://github.com/denoland/deno/blob/master/tools/deno_tcp.ts)!

<img width="696" alt="Zrzut ekranu 2019-08-19 o 16 14 22" src="https://user-images.githubusercontent.com/13602871/63272548-8bf09f80-c29c-11e9-8c02-60c9d3c25b2a.png">

**Comment by lucacasonato** [comment]
Can this be closed? `deno_tcp` averages at about the same latency, or slightly lower than `node_tcp` now.

**Comment by ry** [comment]
Yes - this is fixed.

**Comment by hayd** [comment]
there appears to be a relatively recent (?) regression:

<img width="737" alt="Screen Shot 2021-04-04 at 20 59 00" src="https://user-images.githubusercontent.com/1931852/113520111-cd74b200-9588-11eb-8e25-4093312db2e8.png">

https://deno.land/benchmarks?-100#http-latency

**Comment by AaronO** [comment]
This divergence seems to have started with https://github.com/denoland/deno/commit/fec1b2a5a4324a7eecdfbb2471931f3b6b0139c5 (first commit changing the OP layer)

So this is on my radar, but it's curious that there's a big regression in latency. I'll try to investigate this and correct it for 1.9 (and ideally exceed the original latency)

**Comment by AaronO** [comment]
@hayd I believe I've identified why performance & latency regressed with https://github.com/denoland/deno/commit/fec1b2a5a4324a7eecdfbb2471931f3b6b0139c5.

**TLDR:** `serde_json::Value` with `serde_v8` are very inefficient (take all slow paths)

Ultimately it's a combination of a few factors that should not be present in `1.9`:
- `rumtime/tcp` ops still use serde_json-isms and return `serde_json::Value`s
- These are loosely-typed maps, which take `serde_v8`'s slowpaths
  - Can't use `v8_struct_key` so each field allocates a new v8 string on heap (will cause GC pressure)
  - This happens both for input args and output returns
  - Each key has to go through v8's string write utf8 which converts the strings
- Plus the overhead of allocating the unnecessary `serde_json::Value`

So removing these serde_json-isms like I've done in https://github.com/denoland/deno/pull/10009 should fix the regression and likely deliver better performance.

**Comment by AaronO** [comment]
After further digging, the P99 latency increase appears to be caused by the promiseTable change.

#9843 changed the promiseTable from a JS Object to a Map, which causes higher GC-pressure and thus higher tail latencies.

Checking out the promise ring (https://github.com/denoland/deno/pull/9979), hugely closed the P99 delta between node and deno on the tcp bench

**Comment by AaronO** [comment]
@ry @hayd I would consider this closed with the op improvements that shipped in 1.9.0, `deno_tcp` has lower max latency than `node_tcp` (I still think we should use P99 or something instead of max, max isn't a meaningful statistical measurement and can be caused by a single outlier) current chart:

![image](https://user-images.githubusercontent.com/949223/115144210-b4b9d100-a04b-11eb-8ddd-59540b0f1152.png)
